pubsub/aws/snssqs/snssqs.go

```diff
@@ -0,0 +1,505 @@
+	})
+
+	if err != nil {
+		return "", "", err
+	}
+
+	return *(createTopicResponse.TopicArn), hashedName, nil
+}
+
+// Get the topic ARN from the topics map. If it doesn't exist in the map, try to fetch it from AWS, if it doesn't exist
+// at all, issue a request to create the topic.
+func (s *snsSqs) getOrCreateTopic(topic string) (string, error) {
+	topicArn, ok := s.topics[topic]
+
+	if ok {
+		s.logger.Debugf("Found existing topic ARN for topic %s: %s", topic, topicArn)
+		return topicArn, nil
+	}
+
+	s.logger.Debugf("No topic ARN found for %s\n Creating topic instead.", topic)
+
+	topicArn, hashedName, err := s.createTopic(topic)
+
+	if err != nil {
+		s.logger.Errorf("Error creating new topic %s: %v", topic, err)
+		return "", err
+	}
+
+	// Record topic ARN
+	s.topics[topic] = topicArn
+	s.topicHash[hashedName] = topic
+
+	return topicArn, nil
+}
+
+func (s *snsSqs) createQueue(queueName string) (*sqsQueueInfo, error) {
+	createQueueResponse, err := s.sqsClient.CreateQueue(&sqs.CreateQueueInput{
+		QueueName: aws.String(nameToHash(queueName)),
+		Tags:      map[string]*string{awsSqsQueueNameKey: aws.String(queueName)},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	queueAttributesResponse, err := s.sqsClient.GetQueueAttributes(&sqs.GetQueueAttributesInput{
+		AttributeNames: []*string{aws.String("QueueArn")},
+		QueueUrl:       createQueueResponse.QueueUrl,
+	})
+
+	if err != nil {
+		s.logger.Errorf("Error fetching queue attributes for %s: %v", queueName, err)
+	}
+
+	// Add permissions to allow SNS to send messages to this queue
+	_, err = s.sqsClient.SetQueueAttributes(&(sqs.SetQueueAttributesInput{
+		Attributes: map[string]*string{
+			"Policy": aws.String(fmt.Sprintf(`{
+				"Statement": [{
+					"Effect":"Allow",
+					"Principal":"*",
+					"Action":"sqs:SendMessage",
+					"Resource":"%s"				
+				}]
+			}`, *(queueAttributesResponse.Attributes["QueueArn"]))),
+		},
+		QueueUrl: createQueueResponse.QueueUrl,
+	}))
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &sqsQueueInfo{
+		arn: *(queueAttributesResponse.Attributes["QueueArn"]),
+		url: *(createQueueResponse.QueueUrl),
+	}, nil
+}
+
+func (s *snsSqs) getOrCreateQueue(queueName string) (*sqsQueueInfo, error) {
+	queueArn, ok := s.queues[queueName]
+
+	if ok {
+		s.logger.Debugf("Found queue arn for %s: %s", queueName, queueArn)
+		return queueArn, nil
+	}
+	// creating queues is idempotent, the names serve as unique keys among a given region
+	s.logger.Debugf("No queue arn found for %s\nCreating queue", queueName)
+
+	queueInfo, err := s.createQueue(queueName)
+
+	if err != nil {
+		s.logger.Errorf("Error creating queue %s: %v", queueName, err)
+		return nil, err
+	}
+
+	s.queues[queueName] = queueInfo
+
+	return queueInfo, nil
+}
+
+func (s *snsSqs) Publish(req *pubsub.PublishRequest) error {
+	topicArn, err := s.getOrCreateTopic(req.Topic)
+
+	if err != nil {
+		s.logger.Errorf("Error getting topic ARN for %s: %v", req.Topic, err)
```

Comment:
as convention, please start the log using lowercase. "error getting..."

Reply:
Hmm, I didn't know this was a convention for comments, I'll make another pass